hera_cal/datacontainer.py

```diff
@@ -540,67 +558,84 @@ def __init__(self, data, reds=None, antpos=None, bl_error_tol=1.0):
-
-    def build_red_keys(self, reds):
+
+    def build_red_keys(self, reds: RedundantGroups | list[list[Baseline]]):
         '''Build the dictionaries that map baselines to redundant keys.
 
         Arguments:
-            reds: list of lists of redundant baseline tuples, e.g. (ind1, ind2, pol).
+            reds: list of lists of redundant baselin1e tuples, e.g. (ind1, ind2, pol).
         '''
-        # Map all redundant keys to the same underlying data
-        self.reds = []
-        self._bl_to_red_key = {}
-        self._red_key_to_bls = {}
-        for red in copy.deepcopy(reds):
-            bls_in_data = [bl for bl in red if self.has_key(bl)]
-            if len(bls_in_data) > 1:
-                raise ValueError('RedDataContainer can only be constructed with (at most) one baseline per group, '
-                                 + f'but this data has the following redundant baselines: {bls_in_data}')
-            if len(bls_in_data) == 0:
-                self._add_red(red[0], red)
-            elif len(bls_in_data) > 0:
-                self._add_red(bls_in_data[0], red)
+        
+        if isinstance(reds, RedundantGroups):
+            self.reds = reds
+        else:
+            self.reds = RedundantGroups(red_list=reds, antpos=getattr(self, 'antpos', None))
+                
+        self._data_reds = self.reds.filtered(bls=self.bls())
+        self._reds_keyed_on_data = self.reds.keyed_on_bls(bls=self.bls())
+        
+        # Check that the data only has one baseline per redundant group
+        for red in self._data_reds:
+            if len(red) > 1:        
+                raise ValueError(
+                    'RedDataContainer can only be constructed with (at most) one baseline per group, '
+                    f'but this data has the following redundant baselines: {red}'
+                )
 
         # delete unused data to avoid leaking memory
-        del self[[k for k in self._data if k not in self._bl_to_red_key]]
-
-    def get_ubl_key(self, key):
-        '''Returns the key used interally denote the data stored. Useful for del'''
-        return self._bl_to_red_key[key]
+        del self[[k for k in self._data if k not in self.reds]]
 
+    def get_ubl_key(self, bl):
+        '''Returns the blkey used to internally denote the data stored.
+        
+        If this bl is in a redundant group present in the data, this will return the
+        blkey that exists in the data. Otherwise, it will return the array-wide blkey
+        representing this group.
+        '''
+        return self._reds_keyed_on_data.get_ubl_key(bl)
+        
     def get_red(self, key):
-        '''Returns the list of baselines redundant with this key.'''
-        return self._red_key_to_bls[self._bl_to_red_key[key]]
+        '''Returns the list of baselines in the array redundant with this key.
+        
+        Note: this is not just baselines existing in the data itself, but in the 
+              entire array.
```

Comment:
I'm not sure that this is the behavior we want... is this still true? I could imagine this breaking existing code.

Reply:
I'm fairly confident this was the behaviour of the existing code.